### PR TITLE
Federation

### PR DESCRIPTION
**Pull request by dessalines**
**Update:** This is finally ready to merge into master (or almost). Things we need to check:

- [x] federation tests are passing
- [x] `cargo test` is passing
- [x] if federation is disabled in the config, all federation code is disabled in the backend 
- [x] backend works fine with federation option disabled
- [x] database migration from previous release to this version works fine
- [x] frontend works fine with federation disabled
- [x] Mentions with federated syntax working.

We should redo these same checks before our next stable release, to make sure that the upgrade doesnt break anything.

#145 #578

### CODE REVIEW

File: server/src/routes/federation.rs [hunk 1]
```diff
@@ -1,18 +1,35 @@
...
-    .route(
-      "/federation/u/{user_name}",
-      web::get().to(apub::user::get_apub_user),
-    );
+  if Settings::get().federation.enabled {
+    println!("federation enabled, host is {}", Settings::get().hostname);
+    cfg
+      .service(
+        web::scope("/")
+          .guard(guard::Header("Accept", APUB_JSON_CONTENT_TYPE))
```
Comment (asonix) [inline server/src/routes/federation.rs:15]: you may not want to _only_ handle application/activity+json. Some clients might request application/ld+json, application/json, */*, or nothing at all

Comment (asonix) [reply server/src/routes/federation.rs:15]: it's probably valid to ignore the \*/\* and nothing cases, but you should definitely consider supporting other json-type accept headers

**Comment by Nutomic** [comment]
Updated!

**Comment by dessalines** [comment]
BTW it has to run these advanced_migrations to fill things like the users private and public key columns, and all the actor_ids and ap_ids, and its taking forever, like probably hours for production data. 

I'm gonna have to figure out a way to do this in a background thread so that everything else can still run.

**Comment by Nutomic** [comment]
I think the only thing thats slow with the db migrations is the keypair generation. So it should be enough to run a new thread for each user/community which generates the keypair and saves it in the database. Or however that works in Rust.

**Comment by Nutomic** [comment]
Okay so the federation enabled option looks fine, it just disables all the activitypub related routes so that nothing is going in or out.

I looked through the API code, that looks all good. And I did some basic manual testing of the frontend with federation, nothing broke.

**Comment by dessalines** [comment]
> Unfortunately I cant get the database tests to pass even though postgres is running and cargo run works. Do I need any additional configuration for that?

Try running
```
export LEMMY_DATABASE_URL=postgres://lemmy:password@localhost:5432/lemmy
diesel migration run
```

Its possible your local postgresql doesn't have the new columns.

**Comment by Nutomic** [comment]
Unfortunately that didnt help, getting a bunch of these errors:
`duplicate key value violates unique constraint \"idx_user_name_lower_actor_id\"`
`deadlock detected`

Guess there might be bad data in my DB?

**Comment by dessalines** [comment]
Ya that's likely, I'd clear out your local postgresql DB: `DROP SCHEMA public CASCADE; CREATE SCHEMA public;`

**Comment by dessalines** [comment]
Okay this is ready. This still scares me a lot, but I tested with production data, and everything works.

Mentions look a little different `@user@domain`, but they work both for the future federation, and locally too.

@Nutomic if you would do one last round of testing locally before I merge.

**Comment by Nutomic** [comment]
Federation tests are all passing.

**Pull request merged**